### ISSUE TITLE
Results do not match other implementations

### ISSUE THREAD

**Issue by cburgmer** [codeniko/JsonPathKt#4]
The following queries provide results that do not match those of other implementations of JSONPath
(compare https://cburgmer.github.io/json-path-comparison/):

- [x] `$[:]`
  Input:
  ```
  ["first", "second"]
  ```
  Expected output:
  ```
  ["first", "second"]
  ```
  Error:
  ```
  [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.6.0:java (default-cli) on project query: An exception occured while executing the Java class. Index 0 out of bounds for length 0 -> [Help 1]
  org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.6.0:java (default-cli) on project query: An exception occured while executing the Java class. Index 0 out of bounds for length 0
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
      at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
      at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
      at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
      at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
      at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
      at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke (Method.java:566)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
      at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
      at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
  Caused by: org.apache.maven.plugin.MojoExecutionException: An exception occured while executing the Java class. Index 0 out of bounds for length 0
      at org.codehaus.mojo.exec.ExecJavaMojo.execute (ExecJavaMojo.java:339)
      at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
      at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
      at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
      at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
      at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
      at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
      at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke (Method.java:566)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
      at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
      at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
  Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
      at jdk.internal.util.Preconditions.outOfBounds (Preconditions.java:64)
      at jdk.internal.util.Preconditions.outOfBoundsCheckIndex (Preconditions.java:70)
      at jdk.internal.util.Preconditions.checkIndex (Preconditions.java:248)
      at java.util.Objects.checkIndex (Objects.java:372)
      at java.util.ArrayList.get (ArrayList.java:458)
      at com.nfeld.jsonpathlite.PathCompiler.compileBracket$json_path_lite (PathCompiler.kt:232)
      at com.nfeld.jsonpathlite.PathCompiler.compile$json_path_lite (PathCompiler.kt:59)
      at com.nfeld.jsonpathlite.JsonPath.<init> (JsonPath.kt:17)
      at com.nfeld.jsonpathlite.extension.JSONArrayKt.read (JSONArray.kt:8)
      at com.nfeld.jsonpathlite.JsonArray.read (JsonResult.kt:12)
      at query.AppKt.main (App.kt:10)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke (Method.java:566)
      at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
      at java.lang.Thread.run (Thread.java:834)
  [ERROR] 
  [ERROR] Re-run Maven using the -X switch to enable full debug logging.
  [ERROR] 
  [ERROR] For more information about the errors and possible solutions, please read the following articles:
  [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
  ```

- [x] `$['key','another']`
  Input:
  ```
  {"another": "entry", "key": "value"}
  ```
  Expected output:
  ```
  ["value", "entry"]
  ```
  Actual output:
  ```
  {"key": "value", "another": "entry"}
  ```

- [x] `$.store..price`
  Input:
  ```
  {"store": {"book": [{"category": "reference", "price": 8.95, "title": "Sayings of the Century", "author": "Nigel Rees"}, {"category": "fiction", "price": 12.99, "title": "Sword of Honour", "author": "Evelyn Waugh"}, {"category": "fiction", "price": 8.99, "title": "Moby Dick", "isbn": "0-553-21311-3", "author": "Herman Melville"}, {"category": "fiction", "price": 22.99, "title": "The Lord of the Rings", "isbn": "0-395-19395-8", "author": "J. R. R. Tolkien"}], "bicycle": {"color": "red", "price": 19.95}}}
  ```
  Expected output:
  ```
  [8.95, 12.99, 8.99, 22.99, 19.95]
  ```
  Actual output:
  ```
  [19.95, 8.95, 12.99, 8.99, 22.99]
  ```

For reference, the output was generated by the program in https://github.com/cburgmer/json-path-comparison/tree/master/implementations/Kotlin_com.nfeld.jsonpathlite.

**Comment by codeniko** [comment]
Thanks for this, I really like the page you have showing the differences at https://cburgmer.github.io/json-path-comparison/

I'll take a look at some of these and others in your table over the next several days.

**Comment by codeniko** [comment]
I took a look at the ones you listed in this issue.
1) The first one `$[:]` looks like a crash which I'll fix. I may as well support it too. 

2/3) The second and third ones are based on implementation. I use org.json which intentionally uses a hashmap because the JSON spec says object key order cannot be guaranteed and should not be relied upon. I think Jackson by default also doesn't guarantee order here. With that said, I don't believe the third one is an issue otherwise I'd have to ditch org.json and go with a different parser 🙁 I saw you have an issue created for yourself asking if order is guaranteed on a similar path of `$..key` for jsurfer, but mine happened to be match others.

2) For the second one in particular, I'll have to say that I disagree with the consensus result given and prefer my output. Being that JSON object key order isn't guaranteed, a path like `$['key','another']` will give values in either order and I'd actually be removing the ability to know which value is paired with which key. If I wanted the values themselves similar to the expected consensus output, I can do the path like `$['key','another'][*]` (the * token which I've yet to implement actually and am now reminded to do so).

**Comment by cburgmer** [comment]
Happy if some of the findings can improve the user experience of your implementation!

On the guaranteed order: Thanks for linking to the issue on the other side, I'm indeed not decided on that myself. The consensus at least makes it easy to keep my opinion out so far :) My best take right now is that a guaranteed order would make sense in cases where the user decides to elaborate on the query and say select the last match of the recursive decent, e.g. `$.store..price[-1]`. Obviously, if I don't guarantee ordering I also can't guarantee the result here.

To be overly nit-picky: apparently JSON does allow a key to come up multiple times (still valid!), so some standards (JWT for example) have explicitly called out that the last appearance of a key is respected where uniqueness is required. And it seems many JSON parsers seem to enforce that, with the result that ordering has an impact there.

Just to make my position clear, I do have personal opinions on what I expect for different queries, but those are just opinions of one user. As for the sake of the comparison project, I want to keep my own opinion out of that if I can. As for issue https://github.com/cburgmer/json-path-comparison/issues/3 however, I do need to understand whether it has an effect on how the comparison is made, as it would constitute a new special case to be handled when diffing results.

A side from all that, feel free to close the bug if you have taken the feedback that's relevant for you. Let me know how you want to continue using that comparison, if it is helpful for you. This bug report was meant as an intro - happy to continue this as a service - but maybe the table over at the comparison is all you need for further improvements.

**Comment by codeniko** [comment]
Thank you for the issue, the 1st and 3rd ones will be addressed in the next release: 2.0.0. I'll be switching to Jackson since it allows me to preserve the order of object keys when the JSON is deserialized.

The second one will be kept as designed. My current implementation is more useful in my opinion because it allows for consistency. If I change to only return the values for `$['key','another']` and the JSON object is missing key `key`, then it will only have 1 value and user won't know if the value is from `key` or `another`. Also, keeping the keys allows me to easily convert the result to data class (also new addition to version 2.0.0). Additionally, if I REALLY want the values, I'm thinking one can add `.*` to the jsonpath to get those values. I will implement this token too.

2.0.0 is in the `dev` branch at the moment. I'm going to go through your project and see what other paths I should tackle to be more consistent with other implementations before I release.

**Comment by codeniko** [comment]
2.0.0 has been released and I've addressed all issues in this ticket, including the 2nd one I originally thought I wouldn't change. Closing
